[expo-go][android] add logging of manifest exception

chrfalch opened the pull request:
# Why

When manifest exceptions happens in Expo Go we aren't starting/loading the bundle in a JS context, meaning we don't log errors to the packager - just displaying the error message to the user:

<img width="300" alt="image" src="https://github.com/user-attachments/assets/9ef95e07-d137-4387-8302-107b71f60f92" />

Resolves ENG-15442 together with #36170

# How

This commit fixes this by introducing a helper class that can log to the packager without a JS context in Expo Go:

- Added `ExponentPackageLogger` class for connecting and logging to the packager. It logs a single message using a static method and then closes the web socket.
- Added calling the log function in the Kernel's `handleError` function
- Exposed some internal props in the ManifestException class to be able to create a plain-text edition of the error
- Added methods to `ExceptionUtils` so that we can create a plain-text version to print out on the console

<img width="907" alt="image" src="https://github.com/user-attachments/assets/03bf7f16-de67-41c7-9ade-818b2ce4b5cb" />

# Test Plan

- Run Expo Go and try to open a running app with SDK-52 against main
- Observe that Expo Go displays and error and that the same error is displayed in the CLI

# Checklist

- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[android][expo-go] add logging of manifest exceptions

When manifest exceptions happens in Expo Go we aren't starting/loading the bundle in a JS context, meaning we don't log errors to the packager - just displaying the error message to the user.

This commit fixes this by introducing a helper class that can log to the packager without a JS context in Expo Go:

- Added `ExponentPackageLogger` class for connecting and logging to the packager. It logs a single message using a static method and then closes the web socket.
- Added calling the log function in the Kernel's `handleError` function
- Exposed some internal props in the ManifestException class to be able to create a plain-text edition of the error
- Added methods to `ExceptionUtils` so that we can create a plain-text version to print out on the console

chrfalch committed:
Minor adjustment

chrfalch committed:
Ran spotless

Kudo commented on apps/expo-go/android/expoview/src/main/java/host/exp/exponent/exceptions/ExceptionUtils.kt:
```diff
@@ -61,13 +61,28 @@ object ExceptionUtils {
 
+  fun exceptionToPlainText(exception: Exception): String {
+    if (exception is ManifestException) {
+      return "${exceptionToErrorHeader(exception)}\n\n${replaceHtml(exception.errorMessage)
+      }\n\nHow to fix this error:\n${replaceHtml(exception.fixInstructions)}"
+    }
```
```suggestion
      return """
${exceptionToErrorHeader(exception)}

${replaceHtml(exception.errorMessage)}

How to fix this error:
${replaceHtml(exception.fixInstructions)}"
""".trimIndent()
    }
```

not sure it that's what you want

Kudo commented on apps/expo-go/android/expoview/src/main/java/host/exp/exponent/exceptions/ExceptionUtils.kt:
```diff
@@ -61,13 +61,28 @@ object ExceptionUtils {
+      ?.replace("<br>", "\n")
+      ?.replace("<b>", "")
+      ?.replace("</b>", "") ?: ""
+  }
```
not sure whether [Html](https://developer.android.com/reference/android/text/Html.html) or [HtmlCompat](https://developer.android.com/reference/androidx/core/text/HtmlCompat) would help to strip html tags

chrfalch replied:
I suggest that we merge and get this PR out and then I'll make another one refactoring `ManifestException` on both Android/iOS (as suggested in #36170)

Kudo commented on apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/ExponentPackageLogger.kt:
```diff
@@ -0,0 +1,118 @@
+  private var pendingMessage: String? = null
+  private var logLevel: LogLevel = LogLevel.INFO
+
+  private fun sendMessage(message: String, logLevel: LogLevel): CompletableFuture<Void> {
```
use kotlin coroutine than the java future

chrfalch replied:
Fixed

chrfalch committed:
[expo-go][android] Refactored to use coroutines

Codereview: Changed to use coroutines in Kotlin instead of java futures

chrfalch committed:
Update apps/expo-go/android/expoview/src/main/java/host/exp/exponent/exceptions/ExceptionUtils.kt

Co-authored-by: Kudo Chien <kudo@expo.dev>

chrfalch committed:
[expo-go][android] small adjustments after code review

Kudo commented on apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/ExponentPackageLogger.kt:
```diff
@@ -0,0 +1,131 @@
+     */
+    fun send(appUrl: String, message: String, logLevel: LogLevel) {
+      val logger = ExponentPackageLogger(appUrl.toUri())
+      CoroutineScope(Dispatchers.IO).launch {
```
lean toward to turn this function as `suspend fun send(...`) and remove the `CoroutineScope(Dispatchers.IO).launch {` here. let the callsite to launch the coroutine and it could be better to cancel this coroutine. for example, we can use the exisiting `kernelScope` coroutine scope + `withContext(Dispatchers.IO)`

chrfalch replied:
Good point!

Kudo reviewed (approved):
leave a nit comment to move coroutine scope out from the logger. the other part looks good. thanks for the great work!

chrfalch committed:
[expo-go][android] codereview

Changed send message function to be a suspend function, and the callsite to use the kernelScope.

Kudo commented on apps/expo-go/android/expoview/src/main/java/host/exp/exponent/kernel/Kernel.kt:
```diff
@@ -903,6 +904,15 @@ class Kernel : KernelInterface() {
       ExceptionUtils.exceptionToCanRetry(exception)
     )
+    if (exception is ManifestException) {
+      kernelScope.launch {
```
may need withContext here or inside the suspend function if you need to execute on background thread. the kernelScope by default running on main thread

chrfalch replied:
Agree! Fixed.

chrfalch replied:
```kotlin
kernelScope.launch {
        withContext(Dispatchers.IO) {
          ExponentPackageLogger.send(
            exception.manifestUrl,
            ExceptionUtils.exceptionToPlainText(exception),
            ExponentPackageLogger.LogLevel.ERROR
          )
        }
      }
```

chrfalch committed:
[expo-go][android] codereview - withContext

Wrapped call to ExponentPackagerLocker in withContext using Dispatchers.IO as per codereview comments.
